### PR TITLE
Define linalg.norm gradient at zero as zero (subgradient convention)

### PR DESCRIPTION
**Pull request by amacati**
The gradient of `jax.numpy.linalg.norm` is currently NaN at 0. This is a common pain-point in optimization problems containing normalizations to unit norms. One example are rotations with axis-angles, which produce a normalized axis and an angle magnitude when converted to quaternions (see e.g. in [scipy.spatial.transform](https://github.com/scipy/scipy/blob/be3f7ee909811e665b8bc63fb88d2559af7988a1/scipy/spatial/transform/_rotation_xp.py#L166-L181)).

 #26248 described the issue, and #27935 experimented with a solution. The attempt was abandoned because there was no consensus on how do handle higher-order derivatives. This PR is another shot at getting subgradients into `linalg.norm`.

## Gradient definition
The gradient of the norm is undefined at 0, and we can basically choose any vector from the unit ball around 0. Here we chose 0. The reason is that the norm is even, 

$$
n(-x) = n(x).
$$

Differentiating both sides and using the chain rule on the inner $-x$, we get 

$$
\nabla n(-x) = -\nabla n(x).
$$

The gradient is thus an odd function, and we expect 

$$
\nabla n(0) = -\nabla n(0)
$$

at the origin, which forces $\nabla n(0) = 0$. 0 is thus the only value at the origin that keeps the gradient odd, and hence the value we adapt in this PR.

## Higher-order derivatives
As far as I can tell, #27935 was abandoned because of issues with higher-order derivatives. The norm's Hessian is

$$
\begin{align*}
\nabla^2 n(x) = \frac{1}{\lVert x \rVert}\left(I - \hat{x}\hat{x}^{\top}\right), \\
\hat{x} = \frac{x}{\lVert x \rVert}.
\end{align*}
$$

Its eigenvalues are $0$ along $\hat{x}$, and $1/\lVert x \rVert$ in the other $n - 1$ directions. Along one direction the norm is linear, $n(t\hat{x}) = t$, so the curvature there is $0$. Sideways the curvature is $1/\lVert x \rVert$, which blows up as $x \to 0$. So the eigenvalues at the origin are

$$\{0, \infty, \dots, \infty }.$$

There is no finite matrix for the Hessian to converge to. Hence we choose `NaN` to expresses that this derivative does not exist. Conveniently, the behavior does not change for anything using second derivatives.

## Implementation and performance
`linalg.norm` is likely used by a large majority of the jax user base, and a meaningful slow-down from the custom jvp would hardly be acceptable. I benchmarked the function by looking at the HLO produced by XLA, and found that the max operation was significantly more performant than a `jnp.where`. It introduces no additional reads of x, and the timings are within measurement noise on my machine.

@jakevdp @mattjj Since there wasn't much movement in #26248 I decided to go forward and pitch this as an actual PR. It would be great to know if there is at least some interest in this, or if this has little chance of making it in.

### CODE REVIEW

**Comment by amacati** [comment]
@mattjj Have you had a chance to look at the proposed convention?

**Comment by amacati** [comment]
Any movement on this @mattjj @jakevdp ?
